lib/seattleflu/db/cli/command/etl/kit.py

```diff
@@ -0,0 +1,629 @@
+
+
+def update_test_strip(db: DatabaseSession, document: dict):
+    """
+    Find identifier that matches the test_strip barcode within *document*.
+    Updates *document* to have both the test_strip barcode and
+    the identifier if found.
+    """
+    strip_barcode = document["test_strip"]
+    strip_identifier = find_identifier(db, strip_barcode)
+
+    document["test_strip"] = {
+        "uuid": None,
+        "barcode": strip_barcode
+    }
+
+    if not strip_identifier:
+        LOG.warning(f"Test strip has unknown barcode «{strip_barcode}»")
+
+    elif strip_identifier.set_name not in expected_identifier_sets["test-strips"]:
+        LOG.warning(f"Test strip barcode found in unexpected identifier set «{strip_identifier.set_name}»")
+
+    else:
+        document["test_strip"] = {
+            "uuid": strip_identifier.uuid,
+            "barcode": strip_identifier.barcode
+        }
+
+
+def upsert_kit_with_sample(db: DatabaseSession,
+                           identifier: str,
+                           sample: dict,
+                           additional_details: dict) -> Any:
+    """
+    Upsert kit by its *identifier* to include link to a sample.
+
+    An existing kit has its *sample_id* updated and the provided
+    *additional_details* are merged (at the top level only into)
+    the existing kit details, if any.
+    """
+    LOG.debug(f"Upserting kit «{identifier}»")
+
+    data = {
+        "identifier": identifier,
+        "sample_id": sample.id,
+        "additional_details": Json(additional_details)
+    }
+
+    if sample.type == 'utm':
+        sample_type = "utm_sample_id"
+    elif sample.type == 'rdt':
+        sample_type = "rdt_sample_id"
+
+    # Look for existing kit
+    kit = find_kit(db, identifier)
+
+    # Nothing found → create
+    if not kit:
+        LOG.info("Creating new kit")
+        status = "created"
+        kit = db.fetch_row(sql.SQL("""
+            insert into warehouse.kit (identifier, {}, details)
+                values(%(identifier)s,
+                       %(sample_id)s,
+                       %(additional_details)s)
+            returning kit_id as id,
+                      identifier,
+                      encounter_id,
+                      {} as sample_id
+            """).format(sql.Identifier(sample_type),
+                        sql.Identifier(sample_type)), data)
+
+    # Found kit → update
+    else:
+        status = "updated"
+        # Warn if kit is already linked to a different sample!
+        if ((kit.rdt_sample_id and kit.utm_sample_id) and
+            (sample.id not in {kit.rdt_sample_id, kit.utm_sample_id})):
+            LOG.warning(f"Kit «{kit.id}» already linked to another " +
+                        f"{data['sample_type']} «{kit[data['sample_type']]}»")
+
+        kit = db.fetch_row(sql.SQL("""
+            update warehouse.kit
+               set {} = %(sample_id)s,
+                   details = coalesce(details, null) || %(additional_details)s
```

Comment:
coalescing to null doesn't make sense, as the whole reason for coalesce is to use alternate values when some value is null. Appending `||` to null will produce null, so that an existing kit won't get updated details.